File: Brainstorming/ai_project_manager/frontend/src/types/project_run.ts
// Types for the V1 /api/projects/{id}/runs endpoints.

export type ProjectRunStatus =
  | 'pending'
  | 'running'
  | 'succeeded'
  | 'failed';

export type ProjectRunStepRole = 'orchestrator' | 'dev' | 'qa';

export type ProjectRunStepStatus =
  | 'running'
  | 'succeeded'
  | 'failed'
  | 'rejected';

export interface ProjectRun {
  id: string;
  project_id: string;
  status: ProjectRunStatus;
  started_at: string | null;
  finished_at: string | null;
  error: string | null;
  total_tasks: number;
  done_tasks: number;
  blocked_tasks: number;
  created_at: string;
  updated_at: string | null;
}

export interface ProjectRunStep {
  id: string;
  run_id: string;
  item_id: string | null;
  role: ProjectRunStepRole;
  status: ProjectRunStepStatus;
  iteration: number;
  summary: string;
  detail: string | null;
  started_at: string | null;
  finished_at: string | null;
  created_at: string;
}

export interface ProjectRunDetail {
  run: ProjectRun;
  steps: ProjectRunStep[];
}

File: Brainstorming/ai_project_manager/backend/app/models/project_run.py
"""`ProjectRun` domain model (Pydantic).

Represents a single "Lancer l'implémentation" invocation over a project.
One row per run; counters are updated live by the orchestrator so the
frontend can render a progress bar without scanning step rows.
"""

from datetime import datetime
from enum import Enum
from uuid import UUID

from app.models.base_model import IdStatefulBaseModel


class ProjectRunStatus(str, Enum):
    """Workflow status of a run."""

    PENDING = "pending"
    RUNNING = "running"
    SUCCEEDED = "succeeded"
    FAILED = "failed"


class ProjectRun(IdStatefulBaseModel):
    """A single orchestration run."""

    project_id: UUID
    status: ProjectRunStatus = ProjectRunStatus.PENDING
    started_at: datetime | None = None
    finished_at: datetime | None = None
    error: str | None = None
    total_tasks: int = 0
    done_tasks: int = 0
    blocked_tasks: int = 0
